youtube_to_anki/utils.py

```diff
@@ -24,3 +26,27 @@ def process_audio_chunk(audio: AudioSegment, chunk: Dict) -> AudioSegment:
     Indexes an audio segment by the start and end timestamp provided in the transcript chunk.
     """
     return audio[chunk["start"] : chunk["end"]]
+
+def take_screenshots(video: cv2.VideoCapture, chunks: Tuple[Dict]) -> Iterable[ndarray]:
+    """
+    Processes a video, for each interval taking a screenshot (ndarray, colors last) at around the middle of the interval
+    """
+    screenshots = []
+    # 30 ms offset to account for delay until the next frame
+    ms_targets = ((chunk["start"] + chunk["end"]) // 2 - 30 for chunk in chunks)
+    while True:
+        try:
+            ms_target = ms_targets.__next__()
+        except StopIteration:
+            break
+        while video.isOpened():
+            success, frame = video.read()
+            if success:
+                if video.get(cv2.CAP_PROP_POS_MSEC) >= ms_target:
+                    screenshots.append(frame)
+                    break
+            else:
+                break
```

Comment:
Just a small clean up.

Also, is the second break every supposed to occur? Or should we raise a runtime error when no screenshot can be found?

```suggestion
    for chunk in chunks:
        ms_target = (chunk["start"] + chunk["end"]) // 2 - 30
        success, frame = video.read()
        if success:
            if video.get(cv2.CAP_PROP_POS_MSEC) >= ms_target:
                screenshots.append(frame)
                break
        else:
            break
```

Reply:
`success` will be false if there was no frame read ([docs](https://docs.opencv.org/4.x/d8/dfe/classcv_1_1VideoCapture.html#a473055e77dd7faa4d26d686226b292c1)) - this will normally never happen, except for cases where the subtitle timing is buggy/off and the `ms_target` is after the last frame of the video.

In my testing with real youtube videos it never occurred, but you can manually trigger it by setting the `transcript_chunks[-1]` start and end times after the last video frame. 

If it should ever occur, the `zip` function you use to iterate later will silently drop those invalid frames/cards, which is the best possible behavior I guess since at least there will be some output (with screenshots), and the dropped frames would be invalid either way sine their audio would also be invalid or cut off.

------------

I like your code without the iterator, it looks cleaner! I've used it and added the `while video.isOpened():` before the frame reading. Basically we are reading through the entire video (while there are screenshots left), and we don't know how many frames we have to drop until the next target frame comes up (could be 1 or 2 or a few thousand), so we just have to endlessly loop until either we get a good frame or the video ends (or something closes the video, which should not happen and is thus similar to a `True` loop).